### ISSUE TITLE
Endless skeleton when first opening community on mobile (prior to joining) for 3-5 mins

### ISSUE THREAD

**Issue by churik** [status-im/status-react#21660]
# Bug Report

## Problem 
When opening community for the first time, there is no color for chats that did not have any updates (so no new messages). The issue is very visible on open communities (Status and Status Inu)

#### Expected behavior

can see colours in chats, can open empty channels and see that you need to Join to post.

#### Actual behavior

https://github.com/user-attachments/assets/d54cfac9-7190-46ba-a6af-1419431a9fc3

![Image](https://github.com/user-attachments/assets/83d3bd4f-10a1-46cc-b7ce-70d8be396a84)

### Reproduction

1. Go to discover
2. Open  Status or Status Inu
3. Navigate to any chat that has no unread indicator

### Additional Information

- Status version: nightly 21/11/24
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

**Comment by churik** [comment]
**RESOLVED:**

This issue is expanded; 
After navigation to such chat and going back there other community chats can not be opened, navigation is broken

https://github.com/user-attachments/assets/d42f873a-b57e-4f06-9bfc-fbbe82c67979

I prioritized this issue for 2.34

cc @ilmotta

**Comment by ilmotta** [comment]
@churik I experimented with a new profile from 2.32.0 from App Store in iOS.

1. First problem we know, but it's concerning, also for engagement purposes: the Discover Communities screen take too long to load and the UI feedback is poor. Using the magic of video editing a clipped the screencast below. I then tried to press on various channels, but only my first press on the Announcements channel showed the skeleton. All subsequent presses had no effect. Actually, pressing on channels with the notification dot caused them to be marked as read.

https://github.com/user-attachments/assets/8f260120-b065-42cb-a886-4121a56371a5

2. I then logged out and logged in. As you can see from the clock, just 4min later. Immediately all channels were working as expected.

https://github.com/user-attachments/assets/86195c85-d8e1-4b7b-9093-534e333f3ff7

### Take away?

It's inconclusive as I expected, but it appears that logging out and in can fix the issue sometimes, which implies something in status-go gets stuck, or it could be a client-only bug.

In any case, it seems to be relatively easy to reproduce, so I think there's hope of fixing this.

**Comment by churik** [comment]
@ilmotta Yes, relogging in fixes the issue, but there are multiple entry points where it can reappear—such as any other community, a shared channel you don’t have access to, and each time, it results in broken navigation until you relogin.  

What concerns me the most isn’t just the endless skeleton itself, but rather the **broken navigation** that happens when you go back right after encountering it. I mentioned this [here](https://github.com/status-im/status-mobile/issues/21660#issuecomment-2672028011).  

It’s quite disruptive—chat navigation remains unusable until a relogin.

**Comment by ilmotta** [comment]
@Parveshdhull @flexsurfer this might be a good issue for one of you to eventually take a look for 2.34 due to your experience with navigation. Beyond fixing the immediate bug, as @churik pointed out https://github.com/status-im/status-mobile/issues/21660#issuecomment-2703657725, it would be ideal to make the navigation more resilient. Having to re-login to fix navigation everywhere, means the navigation code is reaching an inconsistent "state" for some reason.

Side note: I understand we may want to rewrite certain parts of navigation, but this boat sailed due to all the priorities set for keycard integration. So the navigation migration will need to be planned for later.

**Comment by ilmotta** [comment]
- Could this issue be linked with https://github.com/status-im/status-mobile/issues/21776 ?

**Comment by churik** [comment]
Yes, good catch, it is also one of the possible  entry points for this navigation problem

**Comment by Parveshdhull** [comment]
Hi @churik, @ilmotta, Thank you for pinging me.

Update from my testing/debugging.

- Discover communities - As you mentioned, @ilmotta, the loading time for discovering communities is quite long. Perhaps we should display a warning or message on the screen while the issue is unresolved. For example: 'Please be patient; it may take 2-3 minutes to load the communities.' Currently, users might ignore this screen after just a few seconds.
- Channel color - After the community is fetched, it takes some extra time to load the community chats. In the meantime, we display channel avatars without a background color. Perhaps we can add a fallback blue color for now, or should we leave it as is? On a related note, it seems that channels are fetching fine now., so probably we can close https://github.com/status-im/status-mobile/issues/21776
- Endless skeleton - We can only remove the skeleton once the chat data is fetched. For some chats, it takes a few seconds, but for certain channels (like the Market channel in Status-Inu), it might take more. Not sure why
- Navigation - From my testing, the navigation worked fine. Yes, after initially opening the Status community, some channels were unresponsive, but they started working after a few seconds. We have [one condition](https://github.com/status-im/status-mobile/blob/110f7c94da0fc04cbf15226057779f57c5e4bb54/src/status_im/contexts/communities/overview/view.cljs#L185) regarding the on-press event: don't allow it unless the community is joined or spectated. So, this delay seems related to the time taken for community spectating. I’m attaching the resulting video, other channels/chat opened fine for me (even after this endless skeleton). I’m unable to fetch the Status community in the debug build, so this observation comes from the PR build and code review. To fix this, we could update the database to set spectated: true as soon as the community is opened, and not wait for the status-go response.

Video: https://github.com/user-attachments/assets/ce362962-e4ec-447b-b51f-2b3618fcfa9c

Actionable Items, if we agree
- [ ] Add fallback background color to channels
- [ ] Set spectated true as soon as community is opened

**Comment by Parveshdhull** [comment]
From the [video](https://github.com/status-im/status-mobile/issues/21660#issuecomment-2680525700) you shared @ilmotta , it do seems navigation is broken. But I am not able to reproduce that scenario.

**Comment by Parveshdhull** [comment]
Upd: I just tried ios, and it seems navigation is breaking there

**Comment by Parveshdhull** [comment]
Found the culprit, navigation issue will be fixed in https://github.com/status-im/status-mobile/pull/22336. (Will add more details there)

**Comment by ilmotta** [comment]
> Navigation - From my testing, the navigation worked fine. Yes, after initially opening the Status community, some channels were unresponsive, but they started working after a few seconds. We have [one condition](https://github.com/status-im/status-mobile/blob/110f7c94da0fc04cbf15226057779f57c5e4bb54/src/status_im/contexts/communities/overview/view.cljs#L185) regarding the on-press event: don't allow it unless the community is joined or spectated. So, this delay seems related to the time taken for community spectating. I’m attaching the resulting video, other channels/chat opened fine for me (even after this endless skeleton). I’m unable to fetch the Status community in the debug build, so this observation comes from the PR build and code review. To fix this, we could update the database to set spectated: true as soon as the community is opened, and not wait for the status-go response.

@Parveshdhull After seeing the video you recorded and the code, I think your suggestion to optimistically update the `:spectated` flag of a community makes sense. The ideal outcome would be to at least register the user's press on the channel and open the chat in the loading state, much better than dropping user presses.

When the handler `:chat.ui/spectate-community` calls the endpoint `wakuext_spectateCommunity`, the `:on-error` handler should revert back the flag `[:communities community-id :spectated]` to false. This will prevent the app from going into an inconsistent state.

A side note @Parveshdhull, the approach we see below should be avoided and it's an anti-pattern, just in case you want to refactor this. It's something Andrea and I extensively discussed and saw causing problems (not here but in other community areas, like assoc'ing calculated permissions into the community map in the app-db). The better approach is to have a separate root state, e.g. `:communities/spectated` indexed by community ID. To give one example, this approach prevents signals with community updates causing race conditions and undoing previous calculations that mutated the community map in the app-db. I hope that makes sense :)

https://github.com/status-im/status-mobile/blob/110f7c94da0fc04cbf15226057779f57c5e4bb54/src/status_im/contexts/communities/events.cljs#L289

**Comment by ilmotta** [comment]
> Channel color - After the community is fetched, it takes some extra time to load the community chats. In the meantime, we display channel avatars without a background color. Perhaps we can add a fallback blue color for now, or should we leave it as is? On a related note, it seems that channels are fetching fine now., so probably we can close https://github.com/status-im/status-mobile/issues/21776

@Parveshdhull I think it's fine to show the channel avatars without background colors until the data arrives. This issue has very little impact. If we pick a default color it would still lead to a possibly confusing scenario where colors switch in front of user's eyes.

**Comment by ilmotta** [comment]
> Discover communities - As you mentioned, @ilmotta, the loading time for discovering communities is quite long. Perhaps we should display a warning or message on the screen while the issue is unresolved. For example: 'Please be patient; it may take 2-3 minutes to load the communities.' Currently, users might ignore this screen after just a few seconds.

The way it works now isn't great for newcomers. Why would a user expect to wait minutes just to see a simple list of communities? The long wait time breaks common sense expectations.

I wonder if we can optimize the endpoint `wakuext_curatedCommunities`, that is, optimize the time it takes for the store node to retrieve a community? Or perhaps fetch communities as soon as the user logs in to _maybe_ save on time? Or centralize more and move to a Status service (like the proxy, but not the proxy necessarily) that could deliver cached results? @Samyoul @jrainville any ideas worth exploring?

The UX suggestion to show users more info while the data hasn't arrived can be helpful @Parveshdhull. Worth bringing to `#design-discussion`.

**Comment by jrainville** [comment]
> Or perhaps fetch communities as soon as the user logs in to maybe save on time?

That would be an option. I think on Dekstop we also fetch only when going to the page, but if we fetch in the background in status-go, it shouldn't be noticeable for users.

> Or centralize more and move to a Status service (like the proxy, but not the proxy necessarily) that could deliver cached results?

I think adding centralization should always be the last solution for real problems. In this case, it is annoying and bad for first time users, but it does work with time and then the community is cached locally so it never happens again.

Actual options would be to stop using store nodes for the community and use actual storage solutions like Codex or even use a contract. There are some posts on the Waku Discuss forums if you want to check it out.

----

In conclusion, there is no perfect solution for now. I think just fetching in the background on app start is the way to go for the time being. The skeleton does show that there is something loading, so adding a "please wait" just repeats information IMO.

**Comment by ilmotta** [comment]
Thank you @jrainville. I would like to understand in more detail why store nodes cannot deliver results faster in the range of less than let's say 30s (instead of minutes) which would be a significant improvement. Perhaps the store node solution reached its peak performance already? I'll check again the posts in Waku Discuss, thanks for the tip.

About the UX, it seems to me that a loading skeleton is a pattern that tells a user something is about to happen soon. When communities take minutes to arrive, loading skeletons lose their effectiveness. We will see what designers can come up with to help us out.

**Comment by jrainville** [comment]
> Thank you @jrainville. I would like to understand in more detail why store nodes cannot deliver results faster in the range of less than let's say 30s (instead of minutes) which would be a significant improvement.

Good question.I just tested a new account on Desktop, and the first two communities loaded in less than 10 seconds.
The Status community loaded in 30-60 seconds, which is too long for sure.

I think the reason the Status community is slower to load is because it is bigger and hits the "max message size", so it needs to be fragmented, hence it's slower to fetch it, because you need to get multiple chunks before you can assemble the community.
Also, the control node is node online right now, so the latest description might be a few days old.

I think if we had a store node solution where "persistent" messages like community descriptions just kept the latest instead of keep all the historic messages, it would be faster too, but then we would need to make the store nodes less agnostic which is why it was not pushed forward by the Waku people.

**Comment by Parveshdhull** [comment]
Removing my assignment as navigation issue is fixed in https://github.com/status-im/status-mobile/pull/22346

I just checked the mobile 2.32.2 release build, and the first two communities were fetched in under 5 seconds (tested three times: under 5, 5, and 15 seconds).

**Comment by ilmotta** [comment]
> Removing my assignment as navigation issue is fixed in [#22346](https://github.com/status-im/status-mobile/pull/22346)
> 
> I just checked the mobile 2.32.2 release build, and the first two communities were fetched in under 5 seconds (tested three times: under 5, 5, and 15 seconds).

This goes as far as to demonstrate how unpredictable/unreliable the experience can be. Thanks for verifying this user journey @Parveshdhull

**Comment by churik** [comment]
Agree that navigation is fixed (hurray @Parveshdhull !), for me it is still around 3 minutes in this state, so I changed the title:

https://github.com/user-attachments/assets/8c8046aa-66dd-4959-b97d-aa1b08809038

The issue with discovery I think was logged [here](https://github.com/status-im/status-mobile/issues/21158)

**Comment by ilmotta** [comment]
@churik, it seems that this issue is being narrowed down to UX issues more than a bug, meaning that improving this can be considered a feature with a user story. This improvement has been captured by the Design issue [Improve how users interact with channels while messages are being fetched in Communities. #22489](https://github.com/status-im/status-mobile/issues/22489)

I'm adding this issue to 2.35 milestone to continue on the steady path of improving the UX around communities.
